### PR TITLE
Java/C#: Re-factor model generator taint tracking configurations to use the new API

### PR DESCRIPTION
**Pull request by michaelnebel**

### CODE REVIEW

**Comment by michaelnebel** [comment]
A couple of remarks for DCA.
The appears to be an increase in cache size about 10%.

For C# DCA reports a minor discrepancy between the generated models.
AFAIK after the new API was introduced some results could be missing when using the "old" taint tracking configuration.
Manually inspecting the reported difference in summaries supports this. The .NET runtime models has not been updated since the new API was introduced and most of the models reported as a difference was generated before the introduction of the new API.
However, there are a couple of extra models being generated and they look sound:
```csv
System.Reflection.Emit;EnumBuilder;false;get_AssemblyQualifiedName;();;Argument[this];ReturnValue;taint;generated
System.Reflection.Emit;EnumBuilder;false;get_FullName;();;Argument[this];ReturnValue;taint;generated
System.Reflection.Emit;TypeBuilder;false;ToString;();;Argument[this];ReturnValue;taint;generated
System.Reflection.Emit;TypeBuilder;false;get_AssemblyQualifiedName;();;Argument[this];ReturnValue;taint;generated
System.Reflection.Emit;TypeBuilder;false;get_FullName;();;Argument[this];ReturnValue;taint;generated
```

File: csharp/ql/src/utils/modelgenerator/internal/CaptureSummaryFlowQuery.qll [hunk 1]
```diff
@@ -79,6 +77,22 @@ private class Activate extends ActiveConfiguration {
  * Captured Model:
  * ```Summaries;BasicFlow;false;AssignToArray;(System.Int32,System.Int32[]);Argument[0];Argument[1].Element;taint```
  */
+module ThroughFlow = TaintTracking::MakeWithState<ThroughFlowConfig>;
```
Comment (aschackmull) [inline csharp/ql/src/utils/modelgenerator/internal/CaptureSummaryFlowQuery.qll:80]: You've moved the qldoc from the predicate below to this module - that looks accidental.

File: csharp/ql/src/utils/modelgenerator/internal/CaptureSinkModelsQuery.qll [hunk 2]
```diff
@@ -0,0 +1,18 @@
+private import semmle.code.csharp.dataflow.ExternalFlow
+private import semmle.code.csharp.dataflow.TaintTracking
+private import CaptureModels
+
+module PropagateToSink = TaintTracking::Make<PropagateToSinkConfig>;
```
Comment (aschackmull) [inline csharp/ql/src/utils/modelgenerator/internal/CaptureSinkModelsQuery.qll:5]: I think I'd favour moving these module instantiations to be next to the configurations. It's completely safe to put multiple such instantiations in scope - unlike the previous configuration classes.

File: java/ql/src/utils/modelgenerator/internal/CaptureSinkModelsQuery.qll [hunk 3]
```diff
@@ -0,0 +1,18 @@
+private import semmle.code.java.dataflow.ExternalFlow
+private import semmle.code.java.dataflow.TaintTracking
+private import CaptureModels
+
+module PropagateToSink = TaintTracking::Make<PropagateToSinkConfig>;
+
+/**
+ * Gets the sink model(s) of `api`, if there is flow from a parameter to an existing known sink.
+ */
+string captureSink(DataFlowTargetApi api) {
```
Comment (aschackmull) [inline java/ql/src/utils/modelgenerator/internal/CaptureSinkModelsQuery.qll:10]: In fact, why not move these predicates back where they were as well?

File: csharp/ql/src/utils/modelgenerator/internal/CaptureSinkModelsQuery.qll (hunk 2 above)
Comment (michaelnebel) [reply csharp/ql/src/utils/modelgenerator/internal/CaptureSinkModelsQuery.qll:5]: Will that cause a performance issue (If one only wants to execute one of the queries then all configuration modules will still be created)? There are three different queries that relies in the content of CaptureModels.qll.

Comment (aschackmull) [reply csharp/ql/src/utils/modelgenerator/internal/CaptureSinkModelsQuery.qll:5]: > Will that cause a performance issue
No.

File: csharp/ql/src/utils/modelgenerator/internal/CaptureSummaryFlowQuery.qll (hunk 1 above)
Comment (michaelnebel) [reply csharp/ql/src/utils/modelgenerator/internal/CaptureSummaryFlowQuery.qll]: Yes. This problem will disappear when moving the module instantiations and predicates.

**Comment by michaelnebel** [comment]
DCA looks similar to the first execution.
Merging.

**Pull request merged**